pkg/pgmodel/ingestor/handler.go

```diff
@@ -72,113 +76,179 @@ func (h *insertHandler) flushPending() {
 	h.pending = NewPendingBuffer()
 }
 
+type labelInfo struct {
+	labelID int32
+	Pos     int32
+}
+
+func labelArrayTranscoder() pgtype.ValueTranscoder { return &pgtype.Int4Array{} }
+
 // Set all seriesIds for a samplesInfo, fetching any missing ones from the DB,
 // and repopulating the cache accordingly.
 // returns: the tableName for the metric being inserted into
 // TODO move up to the rest of insertHandler
 func (h *insertHandler) setSeriesIds(seriesSamples []model.Samples) error {
-	seriesToInsert := make([]model.Samples, 0, len(seriesSamples))
+	seriesToInsert := make([]*model.Series, 0, len(seriesSamples))
 	for i, series := range seriesSamples {
 		if !series.GetSeries().IsSeriesIDSet() {
-			seriesToInsert = append(seriesToInsert, seriesSamples[i])
+			seriesToInsert = append(seriesToInsert, seriesSamples[i].GetSeries())
 		}
 	}
 	if len(seriesToInsert) == 0 {
 		return nil
 	}
 
-	var lastSeenLabel *model.Series
-	batch := h.conn.NewBatch()
-
-	// The epoch will never decrease, so we can check it once at the beginning,
-	// at worst we'll store too small an epoch, which is always safe
-	batch.Queue("BEGIN;")
-	batch.Queue(getEpochSQL)
-	batch.Queue("COMMIT;")
-
-	numSQLFunctionCalls := 0
-	// Sort and remove duplicates. The sort is needed to remove duplicates. Each series is inserted
-	// in a different transaction, thus deadlocks are not an issue.
-	sort.Slice(seriesToInsert, func(i, j int) bool {
-		return seriesToInsert[i].GetSeries().Compare(seriesToInsert[j].GetSeries()) < 0
-	})
-
-	batchSeries := make([][]model.Samples, 0, len(seriesToInsert))
-	// group the seriesToInsert by labels, one slice array per unique labels
-	for _, curr := range seriesToInsert {
-		names, values, ok := curr.GetSeries().NameValues()
-		if !ok {
-			//was already set
-			continue
-		}
-		if lastSeenLabel != nil && lastSeenLabel.Equal(curr.GetSeries()) {
-			batchSeries[len(batchSeries)-1] = append(batchSeries[len(batchSeries)-1], curr)
-			continue
+	metricName := seriesToInsert[0].MetricName()
+	labelMap := make(map[labels.Label]labelInfo, len(seriesToInsert))
+	labelList := model.NewLabelList(len(seriesToInsert))
+	{
+		for _, series := range seriesToInsert {
+			names, values, ok := series.NameValues()
+			if !ok {
+				//was already set
+				continue
+			}
+			for i := range names {
+				key := labels.Label{Name: names[i], Value: values[i]}
+				_, ok = labelMap[key]
+				if !ok {
+					labelMap[key] = labelInfo{}
+					labelList.Add(names[i], values[i])
+				}
+			}
 		}
-
-		batch.Queue("BEGIN;")
-		batch.Queue(getSeriesIDForLabelSQL, curr.GetSeries().MetricName(), names, values)
-		batch.Queue("COMMIT;")
```

Comment:
In my cpu profilings, i did get issues here. Removing these should be very helpful.